### DISCUSSION TITLE
Spatial imputation

### DISCUSSION THREAD

**Discussion by ghost** [thinkingmachines/geowrangler#190] [New Feature Ideas]
I have explored some cases where we need to impute to make sure a grid over a bounding polygon has no missing data inside. Currently `create_zonal_stats` has a `fillna` toggle which can fill all missing cells with 0, but that may not be best to use for all cases. 

I created here a method where a missing cell value is filled with the average of its N neighboring cells with available data. By using `sklearn.neighbors.BallTree`, this does not require the available data cells to be adjacent to the target missing data cell, which is useful when there are large swaths of missing areas.

After imputing, the resulting file looks more spatially consistent compared to when imputed with a constant value (e.g. mean of all cells with available data)

Please see below a sample notebook, let me know what you think!
https://gist.github.com/tm-jace-peralta/f7dc6892e2622eb166ec293936a2872a

**Comment by joshuacortez** [comment]
Nice @tm-jace-peralta  this looks like an elegant solution to spatial imputation! Cool how it doesn't require neighboring grid cells to be adjacent 

How's the runtime? I also wonder if it makes sense to create a version that gets the neighbors purely on the x and y indices instead of the centroid
